lexibank_tls.py

```diff
@@ -617,36 +615,58 @@
 ]
 
 
-class Dataset(BaseDataset):
+@attr.s
+class CustomConcept(pylexibank.Concept):
+    Swahili_Gloss = attr.ib(default=None)
+    Source_ID = attr.ib(default=None)
+
+
+class Dataset(pylexibank.Dataset):
     dir = Path(__file__).parent
     id = "tls"
-    form_spec = FormSpec(
+    concept_class = CustomConcept
+    form_spec = pylexibank.FormSpec(
         brackets={},
         separators=",;/",
         missing_data=("-", "?", "???", "+"),
         replacements=REPLACEMENTS,
     )
 
     def cmd_makecldf(self, args):
+        concepts = {}
+
+        for concept in self.conceptlists[0].concepts.values():
+            idx = concept.id.split("-")[-1] + "_" + slug(concept.english)
+            concepts[concept.number] = idx
+            args.writer.add_concept(
+                ID=idx,
+                Name=concept.english,
+                Concepticon_ID=concept.concepticon_id,
+                Concepticon_Gloss=concept.concepticon_gloss,
+                Swahili_Gloss=concept.attributes["swahili"],
+                Source_ID=concept.number,
+            )
+
         # Add sources
         args.writer.add_sources()
 
         # Add languages
         language_lookup = args.writer.add_languages(lookup_factory="Name")
 
-        # Add concepts
-        concept_lookup = args.writer.add_concepts(
-            id_factory=lambda x: "%s_%s" % (x.id.split("-")[-1], slug(x.gloss)),
-            lookup_factory="Name",
-        )
-
         # TODO: add STEM and PREFIX? (pay attention to multiple forms)
-        for entry in progressbar(self.raw_dir.read_csv("tls.txt", dicts=True)):
+        for entry in pylexibank.progressbar(self.raw_dir.read_csv("tls.txt", dicts=True)):
             # Skip over when language is "Note" (internal comments) or
             # "Gweno1" (a copy of "Gweno")
             if entry["LGABBR"] in ["Note", "Gweno1"]:
                 continue
 
+            if entry["SRCID"].endswith(".0"):
```

Comment:
can't you do that in one turn: `src_idx = entry["SRCID"].strip('.0').strip('.5')`?

Reply:
Well, we still need to substitute the `.5` with `a` for those cases where it matters (as far as I can tell, all numbers ending `.5` correspond to our `a` suffix for the source ID). I'd rather leave this as explicit as possible.

Reply:
Argh, now I get what you mean. Sorry, I'm a bit slow today. :)